File: src/main/java/me/nasukhov/bot/Bot.java
package me.nasukhov.bot;

import me.nasukhov.bot.command.Handler;
import me.nasukhov.bot.io.Channel;
import me.nasukhov.bot.io.ChannelRepository;
import me.nasukhov.bot.io.Input;
import me.nasukhov.bot.io.Output;
import me.nasukhov.bot.task.TaskManager;

import java.util.ArrayList;
import java.util.List;

public class Bot {
    private final List<Handler> handlers = new ArrayList<>();
    private final ChannelRepository channelRepository;

    private final TaskManager taskManager;

    public Bot(ChannelRepository channelRepository, TaskManager taskManager) {
        this.channelRepository = channelRepository;
        this.taskManager = taskManager;
    }

    public void addHandler(Handler handler) {
        handlers.add(handler);
    }

    public void runTasks() {
        taskManager.run();
    }

    public String getName() {
        return "TukitaLangBot";
    }

    // TODO remove output and use output resolver
    public void handle(Input command, Output output) {
        Channel channel = command.channel();
        if (channel.isRegistered()) {
            if (!channel.isActive()) {
                return;
            }
        } else {
            // TODO event driven or something. This domain is a bit leaky
            channelRepository.saveChannel(channel);
            taskManager.registerTasks(channel);
        }

        for (Handler handler : handlers) {
            if (handler.supports(command)) {
                handler.handle(command, output);

                return;
            }
        }
    }
}

File: src/main/kotlin/me/nasukhov/bot/Bot.kt
package me.nasukhov.bot

import me.nasukhov.bot.command.Handler
import me.nasukhov.bot.event.JoinedChannel
import me.nasukhov.bot.event.LeftChannel
import me.nasukhov.bot.event.ReceivedInput
import me.nasukhov.bot.io.Channel
import me.nasukhov.bot.io.Input
import me.nasukhov.bot.io.NoOutput
import me.nasukhov.bot.io.Output
import org.springframework.context.ApplicationEventPublisher
import org.springframework.stereotype.Component

@Component
class Bot(
    initialHandlers: List<Handler>,
    private val eventDispatcher: ApplicationEventPublisher,
) {
    private val handlers: MutableList<Handler> = mutableListOf()

    init {
        initialHandlers.forEach { addHandler(it) }
    }

    fun addHandler(handler: Handler) {
        require(handlers.none { it::class == handler::class }) {
            "More than one handler of the same type is not allowed"
        }

        handlers.add(handler)
    }

    fun joinChannel(channel: Channel) {
        eventDispatcher.publishEvent(JoinedChannel(channel))
    }

    fun leaveChannel(channel: Channel) {
        eventDispatcher.publishEvent(LeftChannel(channel))
    }

    fun handle(command: Input): Output {
        eventDispatcher.publishEvent(ReceivedInput(command))

        for (handler in handlers) {
            if (handler.supports(command)) {
                return handler.handle(command)
            }
        }

        return NoOutput()
    }
}
